src/tests/Stream/SourceFilterTest.cpp

```diff
@@ -0,0 +1,282 @@
+
+/// \brief Test partition filtering.
+///
+
+#include "FlatbufferReader.h"
+#include "Stream/SourceFilter.h"
+#include "WriterModule/template/TemplateWriter.h"
+#include "helpers/SetExtractorModule.h"
+#include <chrono>
+#include <gtest/gtest.h>
+#include <thread>
+#include <trompeloeil.hpp>
+
+class MessageWriterStandIn : public Stream::MessageWriter {
+public:
+  MessageWriterStandIn() : MessageWriter(Metrics::Registrar("", {})) {}
+  MAKE_MOCK1(addMessage, void(Stream::Message const &), override);
+};
+
+class SourceFilterStandIn : public Stream::SourceFilter {
+public:
+  SourceFilterStandIn(Stream::time_point Start, Stream::time_point Stop,
+                      Stream::MessageWriter *Writer, Metrics::Registrar Reg)
+      : SourceFilter(Start, Stop, Writer, Reg) {}
+  using SourceFilter::MessagesDiscarded;
+  using SourceFilter::MessagesReceived;
+  using SourceFilter::MessagesTransmitted;
+  using SourceFilter::RepeatedTimestamp;
+  using SourceFilter::UnorderedTimestamp;
+};
+
+class SourceFilterTest : public ::testing::Test {
+public:
+  Stream::time_point StartTime{std::chrono::system_clock::now()};
+  MessageWriterStandIn Writer;
+  Metrics::Registrar SomeRegistrar{"test_reg", {}};
+  auto getTestFilter() {
+    return std::make_unique<SourceFilterStandIn>(
+        StartTime, std::chrono::system_clock::time_point::max(), &Writer,
+        SomeRegistrar);
+  }
+};
+
+TEST_F(SourceFilterTest, InitState) {
+  auto UnderTest = getTestFilter();
+  EXPECT_FALSE(UnderTest->hasFinished());
+}
+using trompeloeil::_;
+
+class yyyyFbReader : public FileWriter::FlatbufferReader {
+public:
+  bool verify(FileWriter::FlatbufferMessage const &) const override {
+    return true;
+  }
+
+  std::string
+  source_name(FileWriter::FlatbufferMessage const &) const override {
+    return "some_name";
+  }
+
+  uint64_t timestamp(FileWriter::FlatbufferMessage const &) const override {
+    return yyyyFbReader::Timestamp;
+  }
+  static void setTimestamp(uint64_t NewTime) {
+    yyyyFbReader::Timestamp = NewTime;
+  }
+
+private:
+  static uint64_t Timestamp;
+};
+
+uint64_t yyyyFbReader::Timestamp{1};
+
+FileWriter::FlatbufferMessage generateMsg() {
+  std::array<uint8_t, 9> SomeData{'y', 'y', 'y', 'y', 'y', 'y', 'y', 'y', 'y'};
+  setExtractorModule<yyyyFbReader>("yyyy");
+  return FileWriter::FlatbufferMessage(SomeData.data(), SomeData.size());
+}
+
+TEST_F(SourceFilterTest, InvalidMessage) {
+  FORBID_CALL(Writer, addMessage(_));
+  auto UnderTest = getTestFilter();
+  FileWriter::FlatbufferMessage TestMsg;
+  UnderTest->filterMessage(std::move(TestMsg));
+  EXPECT_TRUE(UnderTest->MessagesDiscarded == 1);
+  EXPECT_FALSE(UnderTest->hasFinished());
+}
+
+TEST_F(SourceFilterTest, MessageNoDest) {
+  FORBID_CALL(Writer, addMessage(_));
+  auto UnderTest = getTestFilter();
+  auto TestMsg = generateMsg();
+  UnderTest->filterMessage(std::move(TestMsg));
+  EXPECT_FALSE(UnderTest->hasFinished());
```

Comment:
Considering that the first line of the test says: `FORBID_CALL(Writer, addMessage(_));`. I do not think this is a problem.

Reply:
I saw the `FORBID_CALL`, I mean it's not immediately obvious why you want to enforce that the filter reports not to have finished in this situation. Are you intentionally making sure the filter can't report finished until it has had a destination set, so that it can't finish before it has been set up properly or something?

Reply:
Thats fair, the last line is/was somewhat redundant here.